Add PR testing using pipelines

scottwittenburg opened the pull request:
This PR:

- Adds DAG scheduling (using "needs" instead of "dependencies" in yaml)
- Jobs expose and consume buildcache artifacts more specifically, when enabled
- Removes extra step of pushing generated yaml as a commit, in favor of dynamic child pipelines
- Adds support for embedding `$padding` in `config/install_tree`

In order to re-trigger gitlab pipelines using the correct access token, I needed to close the previous PR (#16027) and open a new one.

eugeneswalker reviewed (approved):
Looks good!

tgamblin commented on lib/spack/spack/ci.py:
```diff
@@ -431,6 +446,15 @@ def pkg_name_from_spec_label(spec_label):
+        deps_set.add(s)
+        for d in s.dependencies(deptype=all_deptypes):
+            deps_set.add(d)
+    return deps_set
```
This method is confusing to me -- if you traverse the concrete spec, you get all the dependencies, so isn't this redundant?

tgamblin commented on lib/spack/spack/ci.py:
```diff
@@ -466,6 +490,11 @@ def generate_gitlab_ci_yaml(env, print_summary, output_file,
+    is_pr_pipeline = False
+    if 'SPACK_IS_PR_PIPELINE' in os.environ:
+        is_pr = os.environ.get('SPACK_IS_PR_PIPELINE').lower()
+        is_pr_pipeline = True if is_pr == 'true' else False
```
You can simplify these four lines by using a default result with `get()`:

```python
is_pr_pipeline = (                                                                                                                                                         
    os.environ.get('SPACK_IS_PR_PIPELINE', '').lower() == 'true'                                                                                                           
)    
```

tgamblin commented on lib/spack/spack/ci.py:
```diff
@@ -601,17 +633,34 @@ def generate_gitlab_ci_yaml(env, print_summary, output_file,
-                        job_dependencies.append(dep_job_name)
+                    if enable_artifacts_buildcache:
+                        transitive_deps = get_all_deps(release_spec)
+                        for trans_dep in transitive_deps:
```
I think you can just traverse directly here rather than calling `get_all_deps` (`traverse()` will do what you want).

tgamblin commented on lib/spack/spack/util/path.py:
```diff
@@ -27,6 +28,35 @@
+SPACK_PATH_PADDING_SIZE = len(SPACK_PATH_PADDING_CHARS)
+
+try:
+    path_max_proc  = subprocess.Popen(['getconf', 'PATH_MAX', '/'],
```
don't do this at module scope -- it means that this is run any time this file is imported (regardless of whether it's used), and enough of these will slow down Spack startup time.   Instead, make functions that get these values.

tgamblin reviewed (changes requested):
@scottwittenburg: LGTM mostly!  I have some minor change requests.

scottwittenburg committed:
Support os-specific $padding in config:install_tree

Providing only $padding or ${padding} results in an attempt to
substitute a padding of maximum system path length, while leaving
room for the parts of the install path spack generates.  Providing
$padding-<len> or ${padding-<len>} simply substitutes padding of
the specified length.

scottwittenburg committed:
Pipelines: Support DAG scheduling and dynamic child pipelines

This change also adds a code path through the spack ci pipelines
infrastructure which supports PR testing on the Spack repository.
Gitlab pipelines run as a result of a PR (either creation or pushing
to a PR branch) will only verify that the packages in the environment
build without error.  When the PR branch is merged to develop,
another pipeline will run which results in the generated binaries
getting pushed to the binary mirror.

scottwittenburg commented:
@tgamblin Thanks for the suggestions, I have addressed the issues you raised and re-written the history into a smaller number of commits.  This could be ready for another round of review or merge.

tgamblin commented on lib/spack/spack/ci.py:
```diff
@@ -601,17 +623,34 @@ def generate_gitlab_ci_yaml(env, print_summary, output_file,
+                    if enable_artifacts_buildcache:
+                        for trans_dep in release_spec.traverse(
+                                deptype=all_deptypes):
+                            if trans_dep is not release_spec:
```
just use:

```
for trans_dep in release_spec.traverse(deptype=all, root=False):
```

And you won't need this if statement.  Note that you can just say "all" without quotes (or with quotes) for the deptypes.  We're repurposing the built-in python `all` function here.

scottwittenburg replied:
Done

tgamblin commented on lib/spack/spack/util/path.py:
```diff
@@ -27,6 +28,38 @@
+
+
+def get_system_path_max():
+    global _spack_sys_max_path_length
```
you could use `@llnl.util.lang.memoized` here to simplify.

scottwittenburg replied:
Well that's cool, thanks for pointing it out 👍

tgamblin commented on lib/spack/spack/ci.py:
```diff
@@ -627,12 +666,31 @@ def generate_gitlab_ci_yaml(env, print_summary, output_file,
+                            # to the jobs "needs", but if artifact buildcache
+                            # is enabled, we'll have to add all transtive deps
+                            # of the compiler as well.
+                            if enable_artifacts_buildcache:
```
It looks like the `enable_artifacts_buildcache` branch is not tested (from the covrerage report).  Can you add a test for this part?

scottwittenburg replied:
Based on the coverage, it looks to me like the branch above is the one that's never tested ( this `if` and the following `else` are never hit).  After looking into it, the reason for that is my test bootstraps `cmake` instead of a compiler (probably because I didn't know how to do what I'm about to ask).

 ~~In order to test these blocks, I think need to find a mock package that can be "built" with a mock compiler, so that eventually this code discovers that a bootstrapped compiler `satisfies` the compiler required to build the spec.  I tried bootstrapping the mock `gcc` and using it as the compiler for some other packages, but so far, that mock `gcc` still doesn't "satisfy" the compiler of the mock package.  Do you know if any mock packages can be tested as built with a mocked compiler?~~

Nevermind that, it seems that `print()` statements in the ci code just weren't making it to the terminal, even if those tests failed.  For some reason, `tty.msg()` statements do get printed though, and now I can see I'm hitting those lines.

scottwittenburg replied:
I added a couple tests to make sure we're generating the right `needs` list for compiler bootstrapping pipelines, when `enable-artifacts-buildcache` is either `True` or `False`.

tgamblin commented on lib/spack/spack/ci.py:
```diff
@@ -601,17 +623,34 @@ def generate_gitlab_ci_yaml(env, print_summary, output_file,
-                            phase_name, strip_compilers, dep_spec, osname,
-                            build_group)
-                        job_dependencies.append(dep_job_name)
+                    if enable_artifacts_buildcache:
```
Is it possible to simplify the two branches here and below?  If not that's fine, but it seems like the logic to add dependencies is duplicated across both sides of the if statement here and in the one down further.  It seems like you could first construct a list of transitive or non-transitive dependencies based on whether the artifacts build cache is enabled, and *then* add the members of the list with a common piece of code.

scottwittenburg replied:
Yep, that was a little yucky.  But I think it's better now.  Let me know what you think.

tgamblin replied:
looks way better!

tgamblin reviewed (changes requested):
@scottwittenburg: a few more requests and then I think this is good to go.

scottwittenburg commented:
@tgamblin Thanks for the suggestions, I'll make another pass.

nithintsk reviewed (commented):
Correction in the commit message for os-specific $padding in config:install_tree.

For padding of a specified length the syntax is `$padding:<len>` and not `$padding-len`. I thought the feature was not working properly until i read the regex.
